cardano-cli/src/Cardano/CLI/Shelley/Run/Key.hs

```diff
@@ -0,0 +1,124 @@
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT,
+                   handleIOExceptT, hoistEither, hoistMaybe, newExceptT)
+
+import           Cardano.Api.TextView (TextViewDescription (..))
+import           Cardano.Api.Typed
+
+import           Cardano.CLI.Byron.Key (ByronKeyFailure, CardanoEra (..),
+                   readEraSigningKey, renderByronKeyFailure)
+import           Cardano.CLI.Helpers
+import           Cardano.CLI.Shelley.Parsers (ITNKeyFile (..), KeyCmd (..),
+                   OutputFile (..), SigningKeyFile (..),
+                   VerificationKeyFile (..))
+
+import qualified Cardano.Crypto.DSIGN.Class as Crypto
+import qualified Cardano.Crypto.Signing as Byron.Crypto
+import qualified Cardano.Crypto.Wallet as Crypto.HD
+
+import qualified Shelley.Spec.Ledger.Keys as Shelley
+
+data ShelleyKeyCmdError
+  = ShelleyKeyCmdReadFileError !(FileError TextEnvelopeError)
+  | ShelleyKeyCmdWriteFileError !(FileError ())
+  | ShelleyKeyCmdByronKeyFailure !ByronKeyFailure
+  | ShelleyKeyCmdByronKeyParseError
+      !Text
+      -- ^ Text representation of the parse error. Unfortunately, the actual
+      -- error type isn't exported.
+  | ShelleyKeyCmdDeserialiseByronVerKeyError
+  | ShelleyKeyCmdItnKeyConvError !ConversionError
+  deriving Show
+
+renderShelleyKeyCmdError :: ShelleyKeyCmdError -> Text
+renderShelleyKeyCmdError err =
+  case err of
+    ShelleyKeyCmdReadFileError fileErr -> Text.pack (displayError fileErr)
+    ShelleyKeyCmdWriteFileError fileErr -> Text.pack (displayError fileErr)
+    ShelleyKeyCmdByronKeyFailure e -> renderByronKeyFailure e
+    ShelleyKeyCmdByronKeyParseError errTxt -> errTxt
+    ShelleyKeyCmdDeserialiseByronVerKeyError ->
+      "Failed to deserialise the provided Byron verification key."
+    ShelleyKeyCmdItnKeyConvError convErr -> renderConversionError convErr
+
+runKeyCmd :: KeyCmd -> ExceptT ShelleyKeyCmdError IO ()
+runKeyCmd cmd =
+  case cmd of
+    KeyConvertByronPaymentKey skfOld skfNew ->
+      runConvertByronPaymentKey skfOld skfNew
+    KeyConvertByronGenesisVerificationKey oldVkf newVkf ->
+      runConvertByronGenesisVerificationKey oldVkf newVkf
+    KeyConvertITNStakeKey itnKeyFile mOutFile ->
+      runSingleITNKeyConversion itnKeyFile mOutFile
+
+runConvertByronPaymentKey
+  :: SigningKeyFile -- ^ Input file: old format
+  -> SigningKeyFile -- ^ Output file: new format
+  -> ExceptT ShelleyKeyCmdError IO ()
+runConvertByronPaymentKey skeyPathOld (SigningKeyFile skeyPathNew) = do
+    sk <- firstExceptT ShelleyKeyCmdByronKeyFailure $
+            readEraSigningKey ByronEra skeyPathOld
+    firstExceptT ShelleyKeyCmdWriteFileError . newExceptT $
+      writeFileTextEnvelope skeyPathNew (Just skeyDesc) (ByronSigningKey sk)
+  where
+    skeyDesc = TextViewDescription "Payment Signing Key"
+
+runConvertByronGenesisVerificationKey
+  :: VerificationKeyFile -- ^ Input file: old format
```

Comment:
Perhaps we should accept this either as a file or as a string CLI arg.

Reply:
this is a rarely used command, we only need to support one